azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -0,0 +1,96 @@
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This is a placeholder policy engine class to test that the regorus
+# dependency is correctly installed.
+# pylint: disable=too-few-public-methods
+
+import pwd
+import os
+import sys
+
+from azurelinuxagent.common import logger
+from azurelinuxagent.common.version import get_distro
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.event import WALAEventOperation, add_event
+from azurelinuxagent.common import conf
+
+def log_policy_info(formatted_string, op=WALAEventOperation.Policy, send_event=True):
+    """Log information to console and telemetry."""
+    logger.info("[Policy] " + formatted_string)
+    if send_event:
+        add_event(op=op, message=formatted_string)
+
+
+class PolicyEngine:
+    """Base class for policy engine"""
+    _instance = None
+
+    @classmethod
+    def get_instance(cls):
+        if cls._instance is None:
+            cls._instance = cls()
+        return cls._instance
+    
+    @staticmethod
+    def _is_policy_supported():
+        distro_info = get_distro()
+        distro_name = distro_info[0]
+        try:
+            distro_version = FlexibleVersion(distro_info[1])
+        except ValueError:
+            return False
+        return distro_name.lower() == 'ubuntu' and distro_version.major >= 16
+
+    def __init__(self):
+        self._initialized = False
+        self._policy_supported = False
+        self._extension_policy_enabled = False
+        self._engine = None
+
+    def initialize(self):
+        try:
+            if self._initialized:
+                return
+            self._policy_supported = self._is_policy_supported()
+
+            if not conf.get_extension_policy_enabled():
+                log_policy_info("Extension policy enforcement is disabled")
+                return
+
+            if not self._policy_supported:
+                log_policy_info("Policy enforcement is not supported on {0}".format(get_distro()))
+                return
+
+            # TO DO - import code is for e2e testing, remove once the binary has been published to GA package
+            user = next((u.pw_name for u in pwd.getpwall() if u.pw_uid == 1000), None)
+            regorus_dir = os.path.join(("/home/" + str(user)), "lib/tests_e2e/tests/executables")
+            sys.path.append(regorus_dir)
+            import regorus  # pylint: disable=import-outside-toplevel
```

Comment:
where is the regorus module defined?

Reply:
when i run the unit tests, this produces a ModuleNotFoundError("No module named 'regorus'")

this exception is being swallowed by the blanket exception handler below and the test pass when they should fail

Reply:
and likely that exception is not reported by the end-to-end tests because it's logged as an INFO, not as an ERROR.

that catch-all and ignore is not a good way to handle those errors

Reply:
I saw your comment below about initialized vs disabled. We should have a sync offline to understand what you intend to handle those errors.

At this point, I see a serious error in this code (importing an undefined module) and that is not being detected by any test. That is concerning.

Reply:
I've intentionally swallowed the import errors, based on our first design discussion where it was mentioned that we should be gracefully handling errors instead of surfacing to the user. If there are import errors, the e2e extension_policy test does fail because _extensions_policy_enabled is still set to false, and the "Extension policy is enabled." message is never logged. I see that I may have misunderstood something, I'll reach out offline to discuss how exactly we want to implement the error handling here.

Reply:
I see the 2 e2e tests passing on your sample run. No problem, we can clarify all of this during our sync. Thanks for your PR!